python/cudf/cudf/core/frame.py

```diff
@@ -3581,64 +3582,148 @@ def _reindex(
         return self._mimic_inplace(result, inplace=inplace)
 
 
-def _get_replacement_values(to_replace, replacement, col_name, column):
+def _get_replacement_values_for_columns(
+    to_replace: Any, value: Any, columns: Dict[Any, Any]
+) -> Tuple[Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]]:
+    """
+    Returns a per column mapping for the values to be replaced, new
+    values to be replaced with and if all the values are empty.
 
-    all_nan = False
-    if (
-        is_dict_like(to_replace)
-        and not isinstance(to_replace, cudf.Series)
-        and replacement is None
+    Parameters
+    ----------
+    to_replace : numeric, str, list-like or dict
+        Contains the values to be replaced.
+    value : numeric, str, list-like, or dict
+        Contains the values to replace `to_replace` with.
+    columns : dict
+        A column to dtype mapping representing dtype of columns.
+
+    Returns
+    -------
+    all_nan_columns : dict
+        A dict mapping of all columns if they contain all na values
+    to_replace_columns : dict
+        A dict mapping of all columns and the existing values that
+        have to be replaced.
+    values_columns : dict
+        A dict mapping of all columns and the corresponding values
+        to be replaced with.
+    """
+    to_replace_columns: Dict[Any, Any] = {}
+    values_columns: Dict[Any, Any] = {}
+    all_nan_columns: Dict[Any, Any] = {}
+
+    if is_scalar(to_replace) and is_scalar(value):
+        to_replace_columns = {col: [to_replace] for col in columns}
+        values_columns = {col: [value] for col in columns}
```

Comment:
It's weird to me that if `to_replace` and `value` are scalars that we wrap them in a list here.

Reply:
The reason I had to wrap these into lists is that our `find_and_replace` expects `to_replace` & `value` to be list-like. Letting any kind of scalar pass through into `find_and_replace` would result in failures, thus doing that prep-processing of ensuring all are list-like / array-likes in this helper where appropriate.